data_model/primitives/src/small.rs

```diff
@@ -0,0 +1,251 @@
+//! Small-string optimisation-related implementations and
+//! structs. Thin wrapper around the `smallstr` crate, using the
+//! `Slavic` conventions. For general use, the array type `[u8; 32]`
+//! should be the standard. It's the best bang-for-the-buck in terms
+//! of stack-based strings.
+#[cfg(not(feature = "std"))]
+use alloc::{borrow::ToOwned, string::String, vec::Vec};
+use core::fmt;
+
+use iroha_schema::{IntoSchema, MetaMap, Metadata};
+use parity_scale_codec::{Decode, Encode, Output};
+use serde::{Deserialize, Deserializer, Serialize};
+use smallstr::SmallString;
+pub use smallvec::{smallvec, Array};
+
+#[derive(Debug, derive_more::Display, Clone, Serialize, Deserialize)]
+/// Wrapper around the [`smallstr::SmallString`] type, enforcing a
+/// specific size of stack-based strings.
+pub struct SmallStr(SmallString<[u8; 32]>);
+
+impl SmallStr {
+    #[must_use]
+    #[inline]
+    /// Construct [`Self`] by taking ownership of a [`String`].
+    pub fn from_string(other: String) -> Self {
+        Self(SmallString::from_string(other))
+    }
+
+    #[must_use]
+    #[inline]
+    #[allow(clippy::should_implement_trait)]
+    /// Construct [`Self`] infallibly without taking ownership of a
+    /// string slice. This is not an implementation of [`FromStr`](core::str::FromStr),
+    /// because the latter implies **fallible** conversion, while this
+    /// particular conversion is **infallible**.
+    pub fn from_str(other: &str) -> Self {
+        Self(SmallString::from_str(other))
+    }
+
+    #[inline]
+    /// Checks if the specified pattern is the prefix of given string.
+    pub fn starts_with(&self, pattern: &str) -> bool {
+        self.0.starts_with(pattern)
+    }
+}
+
+impl<A: Array<Item = u8>> From<SmallString<A>> for SmallStr {
+    fn from(string: SmallString<A>) -> Self {
+        Self(SmallString::from_str(SmallString::as_str(&string)))
+    }
+}
+
+impl<A: smallvec::Array> Extend<A::Item> for SmallVec<A> {
+    fn extend<T: IntoIterator<Item = A::Item>>(&mut self, iter: T) {
+        self.0.extend(iter)
+    }
+}
+
+/// Wrapper struct around [`smallvec::SmallVec`] type. Keeps `N`
+/// elements on the stack if `self.len()` is less than `N`, if not,
+/// produces a heap-allocated vector.
+///
+/// To instantiate a vector with `N` stack elements,
+/// ```ignore
+/// use iroha_data_model::small::SmallVec;
+///
+/// let a: SmallVec<[u8; 24]> = SmallVec(smallvec::smallvec![32]);
+/// ```
+pub struct SmallVec<A: Array>(pub smallvec::SmallVec<A>);
```

Comment:
why didn't you derive `PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize`?

Reply:
The standard: for the same reason I didn't derive `Copy` and `Debug`, because a `SmallVec` of non-`Ord` objects is reasonable, just not `Ord`. 

I know that it's possible to derive `serde` traits for structs with non-(de)serializable fields, but that is only one line shorter than what I did, also requiring that the `SmallVec` only hold types satisfying a certain bound. Same as above, I want `SmallVec` to be (de)serializable only if what I store in it also is, and have the option to store things that are not (de)serializable.

Reply:
yes, sorry, I was aware of this problem, I just didn't see it here